messageID.go

```diff
@@ -11,20 +11,17 @@ type IDGenerator interface {
 // concurrent use by multiple goroutines.
 func NewSafeID(startID int) IDGenerator {
 	return &safeID{
-		nextID: startID,
-		mutex:  &sync.Mutex{},
+		nextID: int64(startID),
 	}
 }
 
 type safeID struct {
-	nextID int
-	mutex  *sync.Mutex
+	nextID int64
 }
 
-func (s *safeID) Next() int {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	id := s.nextID
-	s.nextID++
+func (s *safeID) Next() (id int) {
+	id = int(atomic.LoadInt64(&s.nextID))
+	atomic.AddInt64(&s.nextID, 1)
```

Comment:
for me it looks like a race condition here. Multiple goroutines can load the int...this call is atomic, sure, but could return the same integer if it happens before the first goroutine reached the "AddInt64".

As a prove you can add a test calling the Next() concurrent and process the result via `sort | uniq -c` -> I get some rare (but unwanted) duplicates. ~100 cases in 1mio NextCalls (inside RunParallel)

Reply:
@brainexe will check. thanks for advice.